Gems/Atom/RHI/Code/Source/RHI/Resource.cpp

```diff
@@ -120,4 +120,153 @@ namespace AZ::RHI
         }
         MultiDeviceObject::Shutdown();
     }
+
+    Ptr<ImageView> Resource::GetResourceView(const ImageViewDescriptor& imageViewDescriptor) const
+    {
+        const HashValue64 hash = imageViewDescriptor.GetHash();
+        AZStd::lock_guard<AZStd::mutex> registryLock(m_cacheMutex);
+        auto it = m_resourceViewCache.find(static_cast<uint64_t>(hash));
+        if (it == m_resourceViewCache.end())
+        {
+            return InsertNewImageView(hash, imageViewDescriptor);
+        }
+        else
+        {
+            // We've found a matching ResourceView in the cache, but another thread may be releasing the last intrusive_ptr while
+            // we are in this function, dropping the refcount to 0 (and forcing it to -1 for good measure) then deleting it.
+            //
+            //  There are 2 scenarios:
+            //
+            // m_useCount is -1. The other thread is already on the path to deleting it. We need to make a new one here
+            // and replace the old one.
+            //
+            // m_useCount is >=0. We cannot guarantee another thread won't drop the refcount to 0 after we check the value here
+            // so before we create a new intrusive_ptr, we need to use fetch_add to increment the refcount as we check the value to
+            // prevent a race
+            int useCount = it->second->m_useCount.fetch_add(2);
+            if (useCount == -1)
+            {
+                // The useCount was -1 before we incremented.
+                // Another thread is going to come along and delete the one we just found.
+                // Go ahead and erase it and insert a new one instead
+                m_resourceViewCache.erase(it);
+
+                return InsertNewImageView(hash, imageViewDescriptor);
+            }
+            else
+            {
+                // Create the new Ptr, increasing the refcount
+                Ptr<ImageView> result = static_cast<ImageView*>(it->second);
+
+                // Before we checked the value we artificially incremented the refcount to prevent another
+                // thread from letting it go to 0 again. Get rid of that artificial increase now that we have
+                // our new Ptr to hold on to the refcount
+                it->second->m_useCount.fetch_sub(2);
+                return result;
+            }
```

Comment:
Is this logic copy past from DeviceResource? Can we refactor it so we dont have so many duplictates

Reply:
I do not think there is a way to refactor that without overcomplicating it.